h/static/scripts/controllers/share-widget-controller.js

```diff
@@ -0,0 +1,212 @@
+'use strict';
+
+const Controller = require('../base/controller');
+
+const CONFIG_ATTR = 'share-widget-config';
+const TRIGGER_SELECTOR = `[${CONFIG_ATTR}]`;
+const WIDGET_SELECTOR = '.js-share-widget-owner';
+const TARGET_HREF_ATTR = 'share-target-href';
+const TARGET_HREF_SELECTOR = `[${TARGET_HREF_ATTR}]`;
+const CLIPBOARD_INPUT_SELECTOR = '.js-share-widget-clipboard';
+const PRIVATE_MSG_SELECTOR = '.js-share-widget-msg-private';
+const GROUP_MSG_SELECTOR = '.js-share-widget-msg-group';
+
+const ARROW_PADDING_RIGHT =  16;
+const ARROW_PADDING_BOTTOM =  5;
+
+let shareWidgetAttached = false;
+
+// Travel up the nodes parents to find a matching
+// element with the share-widget configuration values
+const _getParentWithConfig = (node) => {
+  let configNode = null;
+  while ((node = node.parentNode)) {
+    if (node === document.body) {
+      break;
+    }
+    if (node.hasAttribute(CONFIG_ATTR)) {
+      configNode = node;
+      break;
+    }
+  }
+  return configNode;
+};
+
+const _getOffset = (el) => {
+  el = el.getBoundingClientRect();
+  return {
+    // adjust for top left of the document
+    left: el.left + window.scrollX,
+    top: el.top + window.scrollY,
+    width: el.width,
+    height: el.height,
+  };
+};
+
+
+class ShareWidget {
+  constructor(containerElement) {
+
+    // we only attach one to the dom since it's a global listener
+    if (shareWidgetAttached) {
+      return;
+    }
+    shareWidgetAttached = true;
+
+    this._container = containerElement;
+    this._widget = this._container.querySelector(WIDGET_SELECTOR);
+    this._widgetVisible = false;
+
+    // on initialize we need to reset container visbility
+    this.hide();
+
+    this._handler = (event) => {
+
+      const target = event.target;
+
+      // do nothing if we are clicking inside of the widget
+      if (this._container.contains(target)) {
+        return;
+      }
+
+      let configNode = target.hasAttribute(CONFIG_ATTR) ? target : null;
+
+      if (configNode || target.matches(`${TRIGGER_SELECTOR} *`)) {
+
+        // try to track down the parent if we are
+        // matched to be inside of one
+        if (!configNode) {
+          configNode = _getParentWithConfig(target);
+        }
+
+        if (configNode) {
+
+          const config = JSON.parse(configNode.getAttribute(CONFIG_ATTR));
+
+          if (this._widgetVisible) {
+            this.hide();
+          } else {
+            this.showForNode(configNode, config);
+          }
+
+          event.preventDefault();
+          event.stopPropagation();
+          event.stopImmediatePropagation();
+          return;
+        }
+      }
+
+      // hide the widget if the click was not handled by
+      // clicking on the triggers or widget itself
+      if (this._widgetVisible) {
+        this.hide();
+      }
+    };
+
+    window.document.body.addEventListener('click', this._handler);
```

Comment:
We usually access the document object via just `document` rather than `window.document`, it is a little terser.